src/daemon/bundle.rs

```diff
@@ -1,33 +1,45 @@
 // Copyright 2019-2023 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
 
-use crate::utils::db::{car_stream::CarHeader, car_util::load_car};
-use anyhow::Context as _;
+use crate::{
+    networks::{ActorBundleInfo, ACTOR_BUNDLES},
+    utils::{db::car_util::load_car, net::http_get},
+};
+use anyhow::ensure;
+use futures::{stream::FuturesUnordered, TryStreamExt};
 use fvm_ipld_blockstore::Blockstore;
+use std::io::Cursor;
+use tracing::warn;
 
-pub async fn load_actor_bundles(db: &impl Blockstore) -> anyhow::Result<CarHeader> {
-    const ERROR_MESSAGE: &str = "Actor bundles assets are not properly downloaded, make sure git-lfs is installed and run `git lfs pull` again. See <https://github.com/git-lfs/git-lfs/blob/main/INSTALLING.md>";
+/// Tries to load the missing actor bundles to the blockstore. If the bundle is
+/// not present, it will be downloaded.
+pub async fn load_actor_bundles(db: &impl Blockstore) -> anyhow::Result<()> {
+    FuturesUnordered::from_iter(
+        ACTOR_BUNDLES
+            .iter()
+            .filter(|bundle| !db.has(&bundle.manifest).unwrap_or(false))
+            .map(
+                |ActorBundleInfo {
+                     manifest: root,
+                     url,
+                     alt_url,
+                 }| async move {
+                    let response = if let Ok(response) = http_get(url).await {
+                        response
+                    } else {
+                        warn!("failed to download bundle from primary URL, trying alternative URL");
+                        http_get(alt_url).await?
+                    };
+                    let bytes = response.bytes().await?;
+                    let header = load_car(db, Cursor::new(bytes)).await?;
+                    ensure!(header.roots.len() == 1);
+                    ensure!(&header.roots[0] == root);
```

Comment:
nit: we could use also `ensure!(header.roots.first() == Some(root));` instead of `ensure!(&header.roots[0] == root);`. This would avoid us the need of testing for length before (and testing length before **or** after would work).

Reply:
I took from another piece of code. This is not exactly the same, as it would fail if there are more roots, thus ensuring the bundles are _exactly_ what we expect them to be.

Reply:
I was not speaking of removing the header roots length test, since we need both yes. Just replacing array access with `.first()`, but no big deal for sure.